packages/basic-crawler/src/internals/basic-crawler.ts

```diff
@@ -989,18 +1000,33 @@ export class BasicCrawler<Context extends CrawlingContext = BasicCrawlingContext
         }
     }
 
-    protected async _handleFailedRequestHandler(crawlingContext: Context, error: Error): Promise<void> {
-        if (this.failedRequestHandler) {
-            await this._tagUserHandlerError(() => this.failedRequestHandler?.(crawlingContext, error));
-            return;
-        }
-
+    protected async _handleFailedRequestHandler(crawlingContext: Context, error: unknown): Promise<void> {
+        // Always log the last error regardless if the user provided a failedRequestHandler
         const { id, url, method, uniqueKey } = crawlingContext.request;
-        const message = error instanceof TimeoutError && !process.env.CRAWLEE_VERBOSE_LOG ? error.message : error.stack;
+        const message = error instanceof TimeoutError && !process.env.CRAWLEE_VERBOSE_LOG
+            ? error.message : (error as any).stack ?? (error as any).message ?? error;
+
         this.log.error(
             `Request failed and reached maximum retries. ${message}`,
             { id, url, method, uniqueKey },
         );
+
+        if (this.failedRequestHandler) {
+            await this._tagUserHandlerError(() => this.failedRequestHandler?.(this._augmentContextWithDeprecatedError(crawlingContext, error), error));
+        }
+    }
+
+    protected _augmentContextWithDeprecatedError(context: Context, error: unknown) {
+        Object.defineProperty(context, 'error', {
+            get: () => {
+                // eslint-disable-next-line max-len
+                this.log.deprecated("The 'error' property of the crawling context is deprecated, and it is now passed as the second parameter in 'failedRequestHandler' or 'errorHandler'. Please update your code, as this property will be removed in a future version.");
```

Comment:
maybe better this way? idk
```suggestion
                this.log.deprecated("The 'error' property of the crawling context is deprecated, and it is now passed as the second parameter in 'errorHandler' and 'failedRequestHandler'. Please update your code, as this property will be removed in a future version.");
```